### PR TITLE
Fix `materialize_module` recursively setting its child module

### LINKED ISSUE #13042: `nn.Module` subclasses uncollected by `get_all_subclasses` raise `RecursionError` after materialization
## :bug: Bug
Materialization of a `nn.Module` subclass that is not collected by `get_all_subclasses` sets its materialized module to its child and thus leads to `RecursionError`.

### To Reproduce
Here’s the minimal code for repro:

```python
from pytorch_lightning.demos.boring_classes import BoringModel
from pytorch_lightning.utilities.meta import materialize_module, init_meta_context

with init_meta_context():
    model = BoringModel()  # BoringModel is somehow not in `pl.utilities.meta.get_all_subclasses(torch.nn.Module)`

materialize_module(model)

model.layer
model.layer.layer  # should raise AttributeError, but it doesn't atm
model.layer.layer.layer.layer.layer.layer  # should raise AttributeError, but it doesn't atm
print(model)
```
```
Traceback (most recent call last):
  File “/Users/nitta/work/github.com/PyTorchLightning/pytorch-lightning/bug_meta.py”, line 12, in <module>
    print(model)
  File “/Users/nitta/.miniconda3/envs/dev39/lib/python3.9/site-packages/torch/nn/modules/module.py”, line 1828, in __repr__
    mod_str = repr(module)
  File “/Users/nitta/.miniconda3/envs/dev39/lib/python3.9/site-packages/torch/nn/modules/module.py”, line 1828, in __repr__
    mod_str = repr(module)
  File “/Users/nitta/.miniconda3/envs/dev39/lib/python3.9/site-packages/torch/nn/modules/module.py”, line 1828, in __repr__
    mod_str = repr(module)
  [Previous line repeated 328 more times]
  File “/Users/nitta/.miniconda3/envs/dev39/lib/python3.9/site-packages/torch/nn/modules/module.py”, line 1822, in __repr__
    extra_repr = self.extra_repr()
  File “/Users/nitta/.miniconda3/envs/dev39/lib/python3.9/site-packages/torch/nn/modules/linear.py”, line 106, in extra_repr
    return ‘in_features={}, out_features={}, bias={}’.format(
RecursionError: maximum recursion depth exceeded while getting the str of an object
```

### Expected behavior
No error.

### Environment
Same env as the CI.

- PyTorch Lightning Version (e.g., 1.5.0): master
- PyTorch Version (e.g., 1.10): 1.11.0
- Python version (e.g., 3.9): 3.9
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context

In #12984, we’ve found these two of the standalone test cases fail due to `RecusionError`:
```
FAILED tests/strategies/test_deepspeed_strategy.py::test_deepspeed_with_meta_device
FAILED tests/strategies/test_deepspeed_strategy.py::test_deepspeed_with_meta_device
```
on which the minimal code above is based.

cc @otaj

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
~Part of #12373.~
Fixes #13042 which is part of #12984.

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @otaj @borda

### CODE REVIEW

File: .github/workflows/ci_test-full.yml [hunk 1]
```diff
@@ -136,6 +136,11 @@ jobs:
       run: |
         python requirements/check-avail-extras.py
 
+    # FIXME
+    - name: Test init_meta_context
+      run: |
+        PL_RUN_STANDALONE_TESTS=1 python -m pytest tests/utilities/test_meta.py::test_fit_with_init_meta_context -vv
```
Comment (Borda) [inline .github/workflows/ci_test-full.yml:142]: I gues you can run this locally as it does not require any GPU :rabbit:

Comment (akihironitta) [reply .github/workflows/ci_test-full.yml:142]: Yes, I'm debugging it locally now. I just wanted to check environments in which the bug occurs and also wanted to share the result with other people if needed.

File: pytorch_lightning/utilities/meta.py [hunk 2]
```diff
@@ -186,7 +186,7 @@ def materialize_module(root_module: nn.Module) -> nn.Module:
         if not materialize_fn or isinstance(child, (Sequential, ModuleList, ModuleDict)):
             materialize_module(child)
         else:
-            setattr(child, name, materialize_fn())
+            setattr(root_module, name, materialize_fn())
```
Comment (akihironitta) [inline pytorch_lightning/utilities/meta.py:189]: I can't think of a case when `setattr(child, name, materialize_fn())` is valid here. Is there? cc @tchaton as the main author.

---

For example, when we materialize `BoringModel`, their values are:

```
root_module=BoringModel(...)
child=Linear(...)
name="layer"
materialize_fn()=Linear(...)
```

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

**Review by otaj — APPROVED** [review]

File: tests/utilities/test_meta.py [hunk 3]
```diff
@@ -68,3 +70,17 @@ def test_init_meta_context():
...
+    """Test fitting a model initialized on meta device."""
+    with init_meta_context():
+        model = BoringModel()
+    trainer = Trainer(
+        fast_dev_run=True,
+        enable_progress_bar=False,
+        enable_model_summary=False,
+        enable_checkpointing=False,
+    )
+    trainer.fit(model)
```
Comment (awaelchli) [inline tests/utilities/test_meta.py:86]: Fitting is not necessary. Let's instead target a unit test here (directly on the function / decorator)

**Review by carmocca — APPROVED** [review]

**Pull request merged**